File: 791_custom_sort_string.py
class Solution:
    def customSortString(self, order: str, s: str) -> str:
        result = ""
        order_lookup = {c: True for c in order}
        char_counts: dict[str, int] = {}

        for c in s:
            char_counts[c] = char_counts.get(c, 0) + 1

        for c in order:
            if c in char_counts:
                result += c * char_counts[c]

        for c in s:
            if c not in order_lookup:
                result += c

        return result

File: problems/791-custom-sort-string.ts
function customSortString(order: string, s: string): string {
  let result = ""
  const orderedLookup = new Set(order)
  const sCounts = new Map<string, number>()

  for (const char of s) {
    const count = (sCounts.get(char) ?? 0) + 1
    sCounts.set(char, count)
  }

  for (const char of order) {
    const count = sCounts.get(char)
    if (count !== undefined) {
      result += Array(count).fill(char).join("")
    }
  }

  for (const char of s) {
    if (!orderedLookup.has(char)) {
      result += char
    }
  }

  return result
}
